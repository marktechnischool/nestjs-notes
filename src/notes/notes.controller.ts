import { Body, Controller, Get, Post } from '@nestjs/common';
import { NotesService } from './notes.service';
import { CreateNoteDto } from './dto/create-note.dto';

@Controller('notes')
export class NotesController {
  constructor(private readonly notesService: NotesService) {}

  @Post()
  async create(@Body() body: CreateNoteDto) {
    return this.notesService.create(body.text);
  }

  @Get()
  async findAll() {
    return this.notesService.findAll();
  }
}
