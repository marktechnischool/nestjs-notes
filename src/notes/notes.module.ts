import { Module } from '@nestjs/common';
import { MongooseModule } from '@nestjs/mongoose';
import { NoteSchema } from 'src/schemas/note.schema';
import { NotesController } from './notes.controller';
import { NotesService } from './notes.service';

@Module({
  imports: [MongooseModule.forFeature([{ name: 'Note', schema: NoteSchema }])],
  controllers: [NotesController],
  providers: [NotesService],
  exports: [NotesService],
})
export class NotesModule {}
