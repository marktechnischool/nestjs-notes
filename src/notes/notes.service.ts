import { Injectable } from '@nestjs/common';
import { InjectModel } from '@nestjs/mongoose';
import { Model } from 'mongoose';
import { Note } from 'src/schemas/note.schema';

@Injectable()
export class NotesService {
  constructor(@InjectModel(Note.name) private noteModel: Model<Note>) {}

  async create(text: string): Promise<Note> {
    const createdNote = new this.noteModel({ text });
    return createdNote.save();
  }

  async findAll(): Promise<Note[]> {
    return this.noteModel.find().exec();
  }
}
