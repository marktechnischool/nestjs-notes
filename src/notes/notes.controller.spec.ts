import { Test, TestingModule } from '@nestjs/testing';
import { NotesController } from './notes.controller';
import { MongoMemoryServer } from 'mongodb-memory-server';
import { getModelToken } from '@nestjs/mongoose';
import { Connection, connect, Model } from 'mongoose';
import { Note, NoteSchema } from 'src/schemas/note.schema';
import { NotesService } from './notes.service';

describe('NotesController', () => {
  let controller: NotesController;
  let mongod: MongoMemoryServer;
  let mongoConnection: Connection;
  let noteModel: Model<Note>;

  beforeAll(async () => {
    mongod = await MongoMemoryServer.create();
    const uri = mongod.getUri();
    mongoConnection = (await connect(uri)).connection;
    noteModel = mongoConnection.model(Note.name, NoteSchema);

    const module: TestingModule = await Test.createTestingModule({
      controllers: [NotesController],
      providers: [
        NotesService,
        { provide: getModelToken(Note.name), useValue: noteModel },
      ],
    }).compile();

    controller = module.get<NotesController>(NotesController);
  });

  afterAll(async () => {
    await mongoConnection.dropDatabase();
    await mongoConnection.close();
    await mongod.stop();
  });

  afterEach(async () => {
    const collections = mongoConnection.collections;
    for (const key in collections) {
      const collection = collections[key];
      await collection.deleteMany({});
    }
  });

  it('should be defined', () => {
    expect(controller).toBeDefined();
  });

  describe('create', () => {
    it('should create a note', async () => {
      const note = await controller.create({ text: 'test' });
      expect(note.text).toBe('test');

      const notes = await controller.findAll();
      expect(notes).toHaveLength(1);
      expect(notes[0].text).toBe('test');
    });
  });

  describe('findAll', () => {
    it('should return an empty array', async () => {
      const notes = await controller.findAll();
      expect(notes).toHaveLength(0);
    });

    it('should return an array with one note', async () => {
      await noteModel.create({ text: 'test' });
      await noteModel.create({ text: 'test 2' });

      const notes = await controller.findAll();
      expect(notes).toHaveLength(2);
      expect(notes[0].text).toBe('test');
      expect(notes[1].text).toBe('test 2');
    });

    it('should return an array with one note', async () => {
      await noteModel.create({ text: 'test' });
      await noteModel.create({ text: 'test 2' });

      const notes = await controller.findAll();
      expect(notes.map(({ text }) => ({ text }))).toMatchSnapshot();
    });
  });
});
