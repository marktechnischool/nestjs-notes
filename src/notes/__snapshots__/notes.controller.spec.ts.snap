// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NotesController findAll should return an array with one note 1`] = `
[
  {
    "text": "test",
  },
  {
    "text": "test 2",
  },
]
`;
